nilearn_sandbox/mass_univariate/rpbi.py

```diff
@@ -0,0 +1,851 @@
+            gs_array.sizes = np.zeros((gs_array.n_rows))
+            gs_array.sizes[row_id] = score_size
+            gs_array.n_elts = score_size
+            gs_array.max_elts = score_size
+            self.merge(gs_array)
+        else:  # it fits --> updates (efficient)
+            self.data['row'][self.n_elts:new_n_elts] = row_id
+            self.data['col'][self.n_elts:new_n_elts] = col_idx
+            self.data['data'][self.n_elts:new_n_elts] = row_data[col_idx]
+            self.sizes[row_id] += score_size
+            self.n_elts = new_n_elts
+        return
+
+
+### Parcellation building routines ############################################
+def _ward_fit_transform(all_subjects_data, fit_samples_indices,
+                        connectivity, n_parcels, offset_labels):
+    """Ward clustering algorithm on a subsample and apply to the whole dataset.
+
+    Computes a brain parcellation using Ward's clustering algorithm on some
+    images, then averages the signal within parcels in order to reduce the
+    dimension of the images of the whole dataset.
+    This function is used with Randomized Parcellation Based Inference, so we
+    need to save the labels to further perform the inverse transformation
+    operation. The function therefore needs an offset to be applied on the
+    labels so that they are unique across parcellations.
+
+    Parameters
+    ----------
+    all_subjects_data : array_like, shape=(n_samples, n_voxels)
+      Masked subject images as an array.
+
+    fit_samples_indices : array-like,
+      Indices of the samples used to compute the parcellation.
+
+    connectivity : scipy.sparse.coo_matrix,
+      Graph representing the spatial structure of the images (i.e. connections
+      between voxels).
+
+    n_parcels : int,
+      Number of parcels for the parcellations.
+
+    offset_labels : int,
+      Offset for labels numbering.
+      The purpose is to have different labels in all the parcellations that
+      can be built by multiple calls to the current function.
+
+    Returns
+    -------
+    parcelled_data : numpy.ndarray, shape=(n_samples, n_parcels)
+      Average signal within each parcel for each subject.
+
+    labels : np.ndarray, shape=(n_voxels,)
+      Labels giving the correspondance between voxels and parcels.
+
+    """
+    # XXX: Delayed import is a mega hack which is unfortunately
+    # required. In scipy versions < 0.11, this import ends up
+    # importing matplotlib.pyplot. This sets the matplotlib backend
+    # which causes our matplotlib backend setting code in
+    # nilearn/plotting/__init__.py to have no effect. In environment
+    # without X, e.g. travis-ci, that means the tests will fail with
+    # the usual "TclError: no display name and no $DISPLAY environment
+    # variable". Note this is dependent on the order of import,
+    # whichever comes first has the only shot at setting the
+    # matplotlib backend.
+    from sklearn.cluster import FeatureAgglomeration
```

Comment:
This delayed import may not be required anymore, if I understand the comment correctly.